### PR TITLE
Premovement and Prefiring phases. 

### PR DESCRIPTION
**Pull request by pakfront**
This addresses https://github.com/MegaMek/megamek/issues/3400
This creates two new GamePhases (Premovement and Prefiring) to handle things like revealing hidden units (implemented) and ghost contacts (not implemented) along with a PrephaseDisplay for user interaction.

If you have hidden units, these phases provide a button to reveal. Press Done when done with a unit.
Revealed units will be appear at the beginning of the next phase.

### CODE REVIEW

**Comment by sldfgunslinger2766** [comment]
As requested: NPE Logs. The NPE was probably caused on the hosts end as it was his first time using the interface and the game hung at the end of the hosting players PreMovement Phase iirc. 
[hostlogs.zip](https://github.com/MegaMek/megamek/files/8234630/hostlogs.zip)
[logs.zip](https://github.com/MegaMek/megamek/files/8234633/logs.zip)

File: megamek/src/megamek/client/ui/swing/PrephaseDisplay.java [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+     * appropriate for.
+     *
+     * @author arlith
+     */
+    public enum PrephaseCommand implements PhaseCommand {
+        PREPHASE_NEXT("prephaseNext"),
+        PREPHASE_REVEAL("prephaseReveal"),
+        PREPHASE_CANCEL_REVEAL("prephaseCancelReveal");
+
+        // TODO GHOSTTARGET
```
Comment (NickAragua) [inline megamek/src/megamek/client/ui/swing/PrephaseDisplay.java:70]: ?

File: megamek/src/megamek/client/ui/swing/PrephaseDisplay.java [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+    }
+
+    // buttons
+    protected Map<PrephaseCommand, MegamekButton> buttons;
+
+    // let's keep track of what we're shooting and at what, too
+    private int cen = Entity.NONE; // current entity number
+
+    // is the shift key held?
+    private boolean shiftheld;
```
Comment (NickAragua) [inline megamek/src/megamek/client/ui/swing/PrephaseDisplay.java:111]: Are we using shift at all in this user interaction?

**Review by NickAragua — COMMENTED** [review]
A couple of questions/cleanup issues, but looks good overall.

**Review by NickAragua — APPROVED** [review]
Looking good.

**Comment by sldfgunslinger2766** [comment]
Played another game today using the artifact that was generated before this got merged. No NPE issues, but I had an issue with the UI: The 'Next Unit' tab was greyed out in both pre-phases. Also: hitting 'Done' did not change the focus to center on the next unit and I had to manually select the next unit in the sidebar. Is this working as intended? I looked at the logs and there weren't any errors throughout the whole game. I generated a test game to capture a screenshot of the behavior, which is attached.
image

**Comment by pakfront** [comment]
>  The 'Next Unit' tab was greyed out in both pre-phases. Also: hitting 'Done' did not change the focus to center on the next unit and I had to manually select the next unit in the sidebar. 
I had seen this crop a few times in my own testing but thought I had it fixed it, it works as expected for me in my tests. I guess I have yet to figure out a robust way of finding the next unit. If this crops up again please attach a save game.

**Pull request merged**